addons/test_portal/tests/test_mail_links.py

```diff
@@ -0,0 +1,137 @@
+from werkzeug.test import Client
+from werkzeug.urls import url_encode, url_parse
+from werkzeug.wrappers import BaseResponse
+
+from odoo import http
+from odoo.tests import HttpCase
+from odoo.tools import mute_logger
+
+
+class TestMailAccess(HttpCase):
+    def _cleanup(self, location):
+        # we want the absolute path but not the absolute URL, strip out bullshit
+        # scheme & domain, as well as fragment as Werkzeug's client doesn't
+        # support those (TBF they're a client-only concern)
+        return url_parse(location).replace(scheme='', netloc='', fragment='').to_url()
+
+    def do_login(self, client, body, user):
+        h = html.fromstring(body)
+        f = h.find('.//form[@class="oe_login_form"]')
+        fields = {
+            field.get('name'): field.get('value')
+            for field in f.xpath('.//input[@name] | .//button[@name]')
+            if field.get('value')
+        }
+        fields['login'] = fields['password'] = user
+        r = client.post(f.get('action'), data=fields, environ_base={
+            'REMOTE_HOST': 'localhost',
+            'REMOTE_ADDR': '127.0.0.1',
+        })
+        return r.headers['Location'], self._cleanup(r.headers['Location'])
+
+    def do_flow(self, model, user, *, login, token):
+        params = {}
+        Model = self.env[model]
+        if token and 'access_token' in Model._fields:
+            params = {'access_token': '12345'}
+
+        obj = Model.create(params)
+        c = Client(http.root, BaseResponse)
+        if login:
+            self.authenticate(user, user)
+            c.set_cookie('localhost', key='session_id', value=self.session.sid, httponly=True)
+        else:
+            self.env.cr.flush()
+            self.env.cr.clear()
+        location = url = '/mail/view?' + url_encode({
+            'model': model,
+            'res_id': obj.id,
+            **params
+        })
+        while True:
+            r = c.get(url)
+            self.assertLess(r.status_code, 400, f'{url} = {r.status}')
+            if r.status_code // 100 == 2:
+                if url.startswith('/web/login'):
+                    location, url = self.do_login(c, r.data, user)
+                    continue
+                break
+
+            location = r.headers['Location']
+            redir = self._cleanup(location)
+            assert url != redir, "redirection loop"
+            url = redir
+        return obj, location, r
+
+    @mute_logger(
+        'odoo.addons.base.models.res_users', # on every login ~25 lines
+        'odoo.addons.base.models.ir_model', # on every ACL error ~15 lines
+    )
+    def test_mail_view_portal(self):
+        # (user, login, {(logged, token): final URL}
+        OBJECT_URL = '/web#model={model}&id={id}&active_id={id}&cids=1'
+        PORTAL_URL = '/my/thing?model={model}&res_id={id}{token}'
+        NO_ACCESS = '/web#action=mail.action_discuss'
+        PORTAL_ROOT = '/my' # portal user gets bounced from /web to /my
+        cases = [
```

Comment:
Not sure to understand your test setup completely as it is a bit obscure... but if you refer to demo / admin users, you should probably use HttpCaseWithUserDemo class that ensures demo user exists (in case it has been deleted or we are running tests without demo data, which we try to support).

Reply:
Added explanation for the test table. It's really not that obscure.